### DISCUSSION TITLE
Recommended graceful way to restart / reload Caddy

### DISCUSSION THREAD

**Discussion by keybits** [eriksjolund/podman-caddy-socket-activation#32] [General]
Assuming I've used Example 1.

What is the recommended way to gracefully restart / reload Caddy so that changes to Caddyfile are picked up?

**Comment by eriksjolund** [comment]
Unfortunately, I don't know of any way to reload the configuration. I guess these commands should work


Alternative 1:
`systemctl --user restart caddy.service`
Alternative 2:
`systemctl --user stop caddy.service && systemctl --user start caddy.service`



Regarding reloading I opened the issue 

* https://github.com/caddyserver/caddy/issues/6631

about an error message I got 

```
 Error: sending configuration to instance: performing request: Post "http://127.0.0.1/load": dial fd: unknown network fd
```

Today I added a note about it:

* https://github.com/eriksjolund/podman-caddy-socket-activation/pull/33

**Comment by keybits** [comment]
Thanks for looking into this. Nice to see the discussion on the Caddy repo. I'll keep an eye on that and try exploring a bit more (though it'll likely be after the holiday period before I get back to it).

**Comment by pomology** [comment]
https://caddy.community/t/how-to-reload-caddy-so-that-it-reads-the-etc-caddy-caddyserver-file-changes-and-updates-itself/13114

If running as a non root user, I  suggest: `systemctl --user reload caddy ` if you want a no downtime config reload.

**Reply by eriksjolund** [reply]
> If running as a non root user, I suggest: systemctl --user reload caddy  if you want a no downtime config reload.

I tried to follow the same method but I adapted it for podman by using `ExecReload=/usr/bin/podman exec ...`

So I gave it a try and added
```
[Service]
ExecReload=/usr/bin/podman exec systemd-%N /usr/bin/caddy reload --config /etc/caddy/Caddyfile --force
```
 but reloading failed with the error message

```
   Dec 21 18:18:07 fcos caddy[10785]: Error: sending configuration to instance: performing request: Post "http://localhost:2019/load": dial tcp [::1]:2019: connect: connection refused
```

Here all details:

1. Follow example1 from the main branch as of today (https://github.com/eriksjolund/podman-caddy-socket-activation/tree/fa396bc935d345582553f46f0462ddf5e416f18a/examples/example1)
2. Run command
   ```
   systemctl --user cat caddy.service | grep ExecStart
   ```
   The following output is printed
   ```
   ExecStart=/usr/bin/podman run --name systemd-%N --cidfile=%t/%N.cid --replace --rm --cgroups=split --network none --sdnotify=container -d -v %h/Caddyfile:/etc/caddy/Caddyfile:Z docker.io/library/caddy:2.9.0-beta.3 /usr/bin/caddy run --config /etc/caddy/Caddyfile
   ```
   (I ran the command to get inspiration for what to write in the `ExecReload=`)
3. Edit _~/.config/containers/systemd/caddy.container_ according to this diff
   ```diff
   --- /var/home/test/podman-caddy-socket-activation/examples/example1/caddy.container	2024-12-21 17:51:47.841963319 +0000
   +++ /var/home/test/.config/containers/systemd/caddy.container	2024-12-21 18:17:39.088376898 +0000
   @@ -1,3 +1,6 @@
   +[Service]
   +ExecReload=/usr/bin/podman exec systemd-%N /usr/bin/caddy reload --config /etc/caddy/Caddyfile --force
   +
    [Container]
    Exec=/usr/bin/caddy run --config /etc/caddy/Caddyfile
    Image=docker.io/library/caddy:2.9.0-beta.3
   ```
4. Run command
   ```
   systemctl --user daemon-reload
   ```
5. Run command
   ```
   systemctl --user reload caddy.service
   ```
   The following output is printed
   ```
   Job for caddy.service failed.
   See "systemctl --user status caddy.service" and "journalctl --user -xeu caddy.service" for details.
   ```
6. Run command
   ```
   journalctl --user -xeu caddy.service
   ```
   In the output there are for example these lines:
   ```
   Dec 21 18:18:07 fcos systemd[10246]: Reloading caddy.service...
   ░░ Subject: A reload job for unit UNIT has begun execution
   ░░ Defined-By: systemd
   ░░ Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
   ░░ 
   ░░ A reload job for unit UNIT has begun execution.
   ░░ 
   ░░ The job identifier is 75.
   Dec 21 18:18:07 fcos podman[10785]: 2024-12-21 18:18:07.174396697 +0000 UTC m=+0.132320114 container exec 7dabce61556d95c6f80416f02ffd81d6949268c9423f6b6ff393f9b28ea8bba3 (image=docker.io/library/c>
   Dec 21 18:18:07 fcos caddy[10785]: {"level":"info","ts":1734805087.328309,"msg":"using config from file","file":"/etc/caddy/Caddyfile"}
   Dec 21 18:18:07 fcos caddy[10785]: {"level":"info","ts":1734805087.3312755,"msg":"adapted config to JSON","adapter":"caddyfile"}
   Dec 21 18:18:07 fcos caddy[10785]: Error: sending configuration to instance: performing request: Post "http://localhost:2019/load": dial tcp [::1]:2019: connect: connection refused
   Dec 21 18:18:07 fcos podman[10785]: 2024-12-21 18:18:07.371740606 +0000 UTC m=+0.329663888 container exec_died 7dabce61556d95c6f80416f02ffd81d6949268c9423f6b6ff393f9b28ea8bba3 (image=docker.io/libr>
   Dec 21 18:18:07 fcos systemd[10246]: caddy.service: Control process exited, code=exited, status=1/FAILURE
   ░░ Subject: Unit process exited
   ░░ Defined-By: systemd
   ░░ Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
   ░░ 
   ░░ An ExecReload= process belonging to unit UNIT has exited.
   ░░ 
   ░░ The process' exit code is 'exited' and its exit status is 1.
   Dec 21 18:18:07 fcos systemd[10246]: Reload failed for caddy.service.
   ░░ Subject: A reload job for unit UNIT has finished
   ░░ Defined-By: systemd
   ░░ Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
   ░░ 
   ░░ A reload job for unit UNIT has finished.
   ░░ 
   ░░ The job identifier is 75 and the job result is failed.
   ```

**Reply by eriksjolund** [reply]
I had some more success after I removing the line
```
admin off
```
The command
```
systemctl --user reload caddy.service
```
did not fail, but somehow modifications are not picked up.

**Reply by eriksjolund** [reply]
I tried some more and got it working.
 Now
```
systemctl --user reload caddy.service
```
works.

See
* https://github.com/eriksjolund/podman-caddy-socket-activation/pull/44

that adds more docs to [_Support for reloading the Caddy configuration_](https://github.com/eriksjolund/podman-caddy-socket-activation?tab=readme-ov-file#support-for-reloading-the-caddy-configuration) and [Example 4](examples/example4)



Instead of setting `admin` to a file descriptor number, I set it to a unix socket path.

### Some more tips
#### easy solution: bind-mount directory
It's easier to get it working by bind-mounting the directory where _Caddyfile_ is located.

#### slightly more complicated solution: bind-mount file
If you instead would bind-mount _Caddyfile_ directly (that is bind-mounting a file instead of a directory) then
you need to edit _Caddyfile_ without replacing the file. The inode number for the file must not change.
Note, both emacs and vi replaces the edited file by default.

**Reply by keybits** [reply]
Thanks for following up on this. That's super helpful.
I was about to have another go at setting up a system using this approach.
I'll report back if I have any issues.

**Reply by eriksjolund** [reply]
Today I added 

* https://github.com/eriksjolund/podman-caddy-socket-activation/pull/55

All the examples now support `systemctl --user reload caddy.service`.

I'll close this issue as I think reloading is now working fine.
